File: src/api/article.ts

import request from "../request";
import { ArticleType,ArticlePageResponse } from "../types/article";
export interface ArticleResponse{

 id:number

 title:string

 content:string

 category_id:number

 category_name:string

 status:string

 cover:string

 author:string

 likes:string

 views:string

 create_time:string

 update_time:string

}
// 获取文章列表
export async function getArticleList(params:any):Promise<ArticlePageResponse>{

const res = await request.get(
    "/articles",
    {
        params
    }
)


return {

    list:res.list.map((item:ArticleResponse)=>({

        id:item.id,

        title:item.title,

        content:item.content,

        desc:item.content?.slice(0,20)||"",

        cover:item.cover,

        author:item.author,

        categoryId:item.category_id,

        categoryName:item.category_name,

        status:item.status,

        views:item.views,

        likes:item.likes,

        createTime:item.create_time,

        updateTime:item.update_time

    })),

    total:res.total

}

}
// 文章详情
export async function getArticleDetail(id:number){
      
    const res =
    await request.get<any>(
    `/articles/${id}`
    )

    return {

    id:res.id,

    title:res.title,

    content:res.content,

    cover:res.cover,

    author:res.author,

    categoryId:res.category_id,

    status:res.status,

    views:res.views,

    likes:res.likes,

    createTime:res.create_time,

    updateTime:res.update_time

    }
}
// 创建文章
export function createArtilcle(data:any){
  return request.post(
    "/articles",
    data
  )
}
// 更新文章
export function updateArticleApi(id:number,data:any){
  return request.put(
    `/articles/${id}`,
    data
  )
}
// 删除文章
export function forceDeleteArticleApi(id:number){
  return request.delete(`/articles/${id}/force`)
}

// 批量彻底删除
export function batchDeleteArticle(ids:number[]){
  return request.delete("/articles/batch/force",
    {
      data:{
        ids
      }
    }
  )
}
// 恢复文章
export function restoreArticle(id:number){
  return request.patch(`/articles/${id}/restore`)
}

// 批量恢复文章
export function batchRestoreArticle(ids:number[]){
  return request.patch("/articles/batch/restore",ids)
}


// 修改文章状态
export function updateArticleStatus(id:number,status:string){
  return request.patch(
    `/articles/${id}/status`,
    {
      status
    }
  )
}
// 获取点赞状态api
export async function getLikeStatus(id:number){
  const res = await request.get(`/articles/${id}/like/status`
  )
  return res
}
// 点赞
export async function likeArticle(id:number){
  const res = await request.post(`/articles/${id}/like`
  )

  return res
}
// 取消点赞
export async function unlikeArticle(id:number){
  const res = await request.delete(`/articles/${id}/like`
  )
  return res
}
//获取收藏状态
export async function getFavoriteStatus(id:number){
  const res = await request.get(`/articles/${id}/favorite/status`)
  return res
}
// 收藏文章
export async function favoriteArticle(id:number){
  const res = await request.post(`/articles/${id}/favorite` )
  return res
}
// 取消收藏
export async function unfavoriteArticle(id:number){
  const res = await request.delete(`/articles/${id}/favorite`)
  return res
}
// 增加文章阅读量
export async function increaseArticleView(id:number){
  const res = await request.post(`/articles/${id}/view`)
  return res
}

// 添加文章历史记录
export function addHistory(articleId:number){
  return request.post(`/user/history`,null,
    {
      params:{
        article_id:articleId
      }
    }
  )
}

// 获取阅读历史
export function getHistory(){
  return request.get("/user/history"
  )
}

// 获得回收站文章
export function getTrashArticle(params:any){
  return request.get("/articles/trash",{
    params
  })
}

File: app/routers/article.py
from fastapi import APIRouter,Depends,HTTPException
from app.schemas.article import (ArticleCreate,ArticleResponse,ArticlePageResponse,ArticleUpdate,ArticleStatus,BatchDelete)
from app.models.article import Article
from app.database import get_db
from sqlalchemy.orm import Session,joinedload

router = APIRouter(
  prefix="/articles",
  tags=["文章"]
)

# 获得文章
@router.get("",response_model=ArticlePageResponse)
def get_articles(page:int=1,page_size:int=10,keyword:str|None=None,status:str|None=None,category_id:int|None=None,db:Session=Depends(get_db)):

  query = db.query(Article).options(joinedload(Article.category)).filter(Article.status!="trash")
  if keyword:
    query = query.filter(Article.title.like(f"%{keyword}%"))

  if status:
    query = query.filter(Article.status==status)

  if category_id:
    query = query.filter(Article.category_id==category_id)

  total = query.count()

  articles = query.order_by(Article.create_time.desc()).offset((page-1)*page_size).limit(page_size).all()
  data = []
  for article in articles:

        data.append({

            "id":article.id,

            "title":article.title,

            "content":article.content,

            "cover":article.cover,

            "author":article.author,

            "category_id":article.category_id,

            "category_name":
                article.category.name
                if article.category
                else None,


            "status":article.status,

            "views":article.views,

            "likes":article.likes,

            "create_time":article.create_time,

            "update_time":article.update_time

        })
  return {
      "list":data,
      "total":total
    }

# 新增文章
@router.post("",response_model=ArticleResponse)
def create_article(data:ArticleCreate,db:Session=Depends(get_db)):
  article = Article(
    title=data.title,
    content = data.content,
    cover = data.cover,
    category_id = data.category_id,
    status = data.status
  )
  db.add(article)
  db.commit()
  db.refresh(article)
  return article


# 获得回收站文章
@router.get("/trash",response_model=list[ArticleResponse])
def get_trash_article(db:Session=Depends(get_db)):
  articles = db.query(Article).filter(Article.status=="trash").all()
  return articles

# 获取文章详情
@router.get("/{article_id}",response_model=ArticleResponse)
def get_article_detail(article_id:int,db:Session=Depends(get_db)):
  article = db.query(Article).filter(Article.id==article_id).first()
  if not article:
    raise HTTPException(
      status_code=404,
      detail="文章不存在"
    )
  return article

# 修改文章
@router.put("/{article_id}",response_model=ArticleResponse)
def update_article(data:ArticleUpdate,article_id:int,db:Session=Depends(get_db)):
  article = db.query(Article).filter(Article.id==article_id).first()
  if not article:
    raise HTTPException(
      status_code=404,
      detail="文章不存在"
    )
  article.title = data.title
  article.content = data.content
  article.cover = data.cover
  article.category_id = data.category_id
  article.status = data.status
  db.commit()
  db.refresh(article)
  return article

# 移至回收站
@router.delete("/{article_id}")
def delete_article(article_id:int,db:Session=Depends(get_db)):
  article = db.query(Article).filter(Article.id==article_id).first()
  if not article:
    raise HTTPException(
      status_code=404,
      detail="文章不存在"
    )

  article.status = "trash"
  db.commit()
  return{
    "message":"已移入回收站"
  }


# 恢复文章
@router.patch("/{article_id}/restore")
def restore_article(article_id:int,db:Session=Depends(get_db)):
  article = db.query(Article).filter(Article.id==article_id).first()
  if not article:
    raise HTTPException(
      status_code=404,
      detail="文章不存在"
    )
  article.status=(article.old_status or "draft")
  article.old_status = None
  db.commit()
  return{
    "message":"恢复成功"
  }

# 批量恢复文章
@router.patch("/batch/restore")
def batch_restore(ids:list[int],db:Session=Depends(get_db)):
  articles = db.query(Article).filter(Article.id.in_(ids))
  for article in articles:
    article.status=(article.old_status or "draft")
    article.old_status=None
  db.commit()
  return{
      "message":"批量恢复成功"
    }

# 彻底删除文章
@router.delete("/{article_id}/force")
def force_delete_article(article_id:int,db:Session=Depends(get_db)):
   article = db.query(Article).filter(Article.id==article_id).first()
   if not article:
      raise HTTPException(
        status_code=404,
        detail="文章不存在"
      )
   db.delete(article)
   db.commit()
   return{
     "message":"文章已彻底删除"
   }

# 批量删除文章
@router.delete("/batch/force")
def batch_force_delete(data:BatchDelete,db:Session=Depends(get_db)):
  articles = db.query(Article).filter(Article.id.in_(data.ids))
  for article in articles:
    db.delete(article)
  db.commit()
  return{
    "message":f"成功删除{len(articles)}篇文章"
  }

# 修改文章状态
@router.patch("/{article_id}/status")
def update_article_status(article_id:int,data:ArticleStatus,db:Session=Depends(get_db)):
   article = db.query(Article).filter(Article.id==article_id).first()
   if not article:
      raise HTTPException(
        status_code=404,
        detail="文章不存在"
      )
   
   if data.status=="trash":
      article.old_status = article.status

   article.status = data.status
   db.commit()
   return{
     "message":"文章状态修改成功"
   }



